src/serialization/containers.h

```diff
@@ -0,0 +1,128 @@
+template <template <bool> class Archive, class T> bool do_serialize(Archive<true> &ar, std::vector<T> &v);
+
+template <template <bool> class Archive, class T> bool do_serialize(Archive<false> &ar, std::deque<T> &v);
+template <template <bool> class Archive, class T> bool do_serialize(Archive<true> &ar, std::deque<T> &v);
+
+template<typename K, typename V>
+class serializable_unordered_map: public std::unordered_map<K, V>
+{
+public:
+  typedef typename std::pair<K, V> value_type;
+  typename std::unordered_map<K, V> &parent() { return *this; }
+};
+
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<false> &ar, serializable_unordered_map<K, V> &v);
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<true> &ar, serializable_unordered_map<K, V> &v);
+
+template<typename K, typename V>
+class serializable_map: public std::map<K, V>
+{
+public:
+  typedef typename std::pair<K, V> value_type;
+  typename std::map<K, V> &parent() { return *this; }
+};
+
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<false> &ar, serializable_map<K, V> &v);
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<true> &ar, serializable_map<K, V> &v);
+
+template<typename K, typename V>
+class serializable_unordered_multimap: public std::unordered_multimap<K, V>
+{
+public:
+  typedef typename std::pair<K, V> value_type;
+  typename std::unordered_multimap<K, V> &parent() { return *this; }
+};
+
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<false> &ar, serializable_unordered_multimap<K, V> &v);
+template <template <bool> class Archive, typename K, typename V> bool do_serialize(Archive<true> &ar, serializable_unordered_multimap<K, V> &v);
+
+template <template <bool> class Archive, class T> bool do_serialize(Archive<false> &ar, std::unordered_set<T> &v);
+template <template <bool> class Archive, class T> bool do_serialize(Archive<true> &ar, std::unordered_set<T> &v);
+
+template <template <bool> class Archive, class T> bool do_serialize(Archive<false> &ar, std::set<T> &v);
+template <template <bool> class Archive, class T> bool do_serialize(Archive<true> &ar, std::set<T> &v);
+
+namespace serialization
+{
+  namespace detail
+  {
+    template <typename T> void do_reserve(std::vector<T> &c, size_t N) { c.reserve(N); }
+    template <typename T> void do_add(std::vector<T> &c, T &&e) { c.emplace_back(std::forward<T>(e)); }
+
+    template <typename T> void do_add(std::deque<T> &c, T &&e) { c.emplace_back(std::forward<T>(e)); }
+
+    template <typename K, typename V> void do_add(serializable_unordered_map<K, V> &c, std::pair<K, V> &&e) { c.insert(std::forward<std::pair<K, V>>(e)); }
+
+    template <typename K, typename V> void do_add(serializable_map<K, V> &c, std::pair<K, V> &&e) { c.insert(std::forward<std::pair<K, V>>(e)); }
```

Comment:
I tried (unsuccessfully) to describe this earlier - the usage of `std::forward` here should work, but isn't the intended usage. The usage here should be the equivalent to a `std::move`. Only the `T&&` cases should use `std::forward`. The difference is that `template<typename T> foo(T&& arg)` undergoes ["reference collapsing"](http://thbecker.net/articles/rvalue_references/section_08.html) when `foo` is called, as part of the algorithm to compute `T`. The `const`ness and reference for the argument isn't computed with `K` and `V` here (in this PR).

Example:
```c++
template<typename T> foo(T&&);

int main()
{
  foo(std::string{}); // T -> std::string

  std::string one;
  foo(one); // T -> std::string&

  const std::string two;
  foo(two); // T -> const std::string&
}
```
whereas `std::pair<K, V>&&` is fixed to move/temporaries of type `std::pair` only.

Reply:
Learned about [perfect forwarding](https://eli.thegreenplace.net/2014/perfect-forwarding-and-universal-references-in-c) today. What you say makes sense vtnerd. @moneromooo-monero it seems like you could replace `std::forward->std::move` since `std::forward` will always return an r-value reference (i.e. it will always be passed an r-value reference). An example why an l-value can't be passed into `e`:
```
void print(int &x)
{
	std::cout << x << " l-value print\n";
}

void print(int &&x)
{
	std::cout << x << " r-value print\n";
}

template <typename T>
void example(T &x, T &&y)
{
	std::cout << x << '\n';		//for compiler warning
	print(std::forward<T>(y));
}

template <typename T>
void example2(T &&x)
{
	std::cout << x << '\n';		//for compiler warning
	print(std::forward<T>(x));
}

int main()
{
	int x{3};

	example(x,4);	  //r-value print
	//example(x,x);  //error, no matching function; special type-deduction rule doesn't apply
	example2(x);	  //l-value print, special type-deduction rule applies
	example2(4);	  //r-value print

	return 0;
}
```

Reply:
vtnerd might it be better to keep `std::forward` if it reflects the intended behavior of `do_add`? e.g. if there were an overloaded version that accepted l-values, would they be forwarded or moved?